cartridge/twophase.lua

```diff
@@ -267,27 +267,193 @@ local function reapply(data)
     end
 end
 
+
+--- Execute two-phase commit algorithm on instances.
+-- This function implements two-phase commit algorithm with the following steps:
+--
+-- I. Executes the preparation phase (`opts.fn_prepare`) on every server
+-- (passed as opts.uri_list).
+--
+-- II. If any server reports an error, executes the abort phase (`opts.fn_abort`).
+-- All servers prepared so far are rolled back and unlocked.
+--
+-- III. Performs the commit phase (`opts.fn_commit`).
+-- In case the phase fails, an automatic rollback is impossible, the
+-- cluster should be repaired manually.
+--
+-- @function twophase_commit
+-- @tparam table  opts
+-- @param opts.upload_data
+--   data that will be shared through twophase_commit (any Lua object)
+-- @tparam table  opts.uri_list
+--   array of URIs for performing twophase_commit
+-- @tparam string opts.fn_prepare
+--   name of function that processes prepare phase of thowhase commit
+-- @tparam string opts.fn_commit
+--   name of function that processes commit phase of thowhase commit
+-- @tparam string opts.fn_prepare
+--   name of function that processes abort phase of thowhase commit
+-- @tparam ?string opts.activity_name
+--   name of executing activity with use of twophase commit
+--   (used to prettify log messages)
+--
+-- @treturn[1] boolean true
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+local function twophase_commit(opts)
```

Comment:
To be honest I expected to see constructor that create a function:

```lua
my_2pc = 2pc.new({upload = ..., activity_name = ..., fn = ...})
local ok, err = 2pc({args}, {some possible options = ... })
```

Not sure it's really needed but this way this module will looks like completely standalone library.
cc @rosik

Reply:
After verbal discussion we decided to skip it for now